cudamapper/src/main.cu

```diff
@@ -397,415 +182,341 @@ void align_overlaps(DefaultDeviceAllocator allocator,
     }
-    else
-    {
-        target_parser = io::create_kseq_fasta_parser(parameters.target_filepath, parameters.k + parameters.w - 1);
-    }
-
-    std::cerr << "Query file: " << parameters.query_filepath << ", number of reads: " << query_parser->get_num_seqences() << std::endl;
-    std::cerr << "Target file: " << parameters.target_filepath << ", number of reads: " << target_parser->get_num_seqences() << std::endl;
-
-    // Data structure for holding overlaps to be written out
-    std::mutex overlaps_writer_mtx;
 
-    struct QueryTargetsRange
+    // process device batches one by one
+    for (const IndexBatch& device_batch : batch.device_batches)
     {
-        IndexDescriptor query_range;
-        std::vector<IndexDescriptor> target_ranges;
-    };
-
-    ///Factor of 1000000 to make max cache size in MB
-    std::vector<IndexDescriptor> query_index_descriptors  = group_reads_into_indices(*query_parser,
-                                                                                    parameters.index_size * 1000000);
-    std::vector<IndexDescriptor> target_index_descriptors = group_reads_into_indices(*target_parser,
-                                                                                     parameters.target_index_size * 1000000);
+        process_one_device_batch(device_batch,
+                                 device_cache,
+                                 application_parameters,
+                                 device_allocator,
+                                 overlaps_and_cigars_to_write,
+                                 cuda_stream);
+    }
+}
 
-    //First generate all the ranges independently, then loop over them.
-    std::vector<QueryTargetsRange> query_target_ranges;
+/// \brief does post-processing and writes data to output
+/// \param device_id
+/// \param application_parameters
+/// \param overlaps_and_cigars_to_write new data is added as it gets available
+/// \param output_mutex controls access to output to prevent race conditions
+void writer_thread_function(const int32_t device_id,
+                            const ApplicationParameters& application_parameters,
+                            ThreadsafeProducerConsumer<OverlapsAndCigars>& overlaps_and_cigars_to_write,
+                            std::mutex& output_mutex)
+{
+    CGA_NVTX_RANGE(profiler, ("main::writer_thread_for_device_" + std::to_string(device_id)).c_str());
+    // This function is expected to run in a separate thread so set current device in order to avoid problems
+    CGA_CU_CHECK_ERR(cudaSetDevice(device_id));
 
-    int target_idx = 0;
-    for (const IndexDescriptor& query_index_descriptor : query_index_descriptors)
+    // keep processing data as it arrives
+    cga_optional_t<OverlapsAndCigars> data_to_write = overlaps_and_cigars_to_write.get_next_element();
+    while (data_to_write) // if optional is empty that means that there will be no more overlaps to process and the thread can finish
```

Comment:
Optional: I guess it's a matter of taste, but I would find `while(data_to_write = overlaps_and_cigars_to_write.get_next_element())` more intuitive than the assignment before and in the loop. It tells me while this assignment makes sense, do the loop.

Reply:
You're right, that does look more readable.
done

Reply:
:) Glad you agree.